File: storage-sdk/src/main/java/com/tencent/wink/storage/winkdb/viewmodel/UserViewModel.java
package com.tencent.wink.storage.winkdb.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import com.tencent.wink.storage.winkdb.model.User;
import com.tencent.wink.storage.winkdb.relation.UserAndLibrary;
import com.tencent.wink.storage.winkdb.relation.UserWithMusicLists;
import com.tencent.wink.storage.winkdb.repository.UserRepository;

import java.util.List;

/**
 * Created by walkerzpli on 2021/9/22.
 */
public class UserViewModel extends AndroidViewModel {

    private final UserRepository mRepository;
    private final LiveData<List<User>> mAllUsers;
    private final LiveData<List<UserAndLibrary>> mAllUsersAndLibrary;
    private final LiveData<List<UserWithMusicLists>> mAllUsersWithMusicLists;

    public UserViewModel(@NonNull Application application) {
        super(application);
        mRepository = new UserRepository(application);
        mAllUsers = mRepository.getAllUsers();
        mAllUsersAndLibrary = mRepository.getAllUsersAndLibrary();
        mAllUsersWithMusicLists = mRepository.getAllUsersAndMusicLists();
    }

    public LiveData<User> getUser(int id) {
        return mRepository.getUser(id);
    }

    public LiveData<List<User>> getAllUsers() {
        return mAllUsers;
    }

    public LiveData<List<UserAndLibrary>> getAllUsersWithLibrary() {
        return mAllUsersAndLibrary;
    }

    public LiveData<List<UserWithMusicLists>> getAllUsersWithMusicLists() {
        return mAllUsersWithMusicLists;
    }

    public void insert(User user) {
        mRepository.insert(user);
    }


    public void delete(User user) {
        mRepository.delete(user);
    }

}

File: storage-ktx/src/main/java/com/demo/storage_ktx/viewmodel/UserViewModel.kt
package com.demo.storage_ktx.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.demo.storage_ktx.model.User
import com.demo.storage_ktx.repository.UserRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Created by walkerzpli on 2021/9/22.
 */
class UserViewModel(private val mRepository: UserRepository) : ViewModel() {

    private val mAllUsers: LiveData<List<User>> = mRepository.allUsers

    fun getUser(id: Int): LiveData<User> {
        return mRepository.getUser(id)
    }

    val allUsers: LiveData<List<User>>
        get() = mAllUsers

    fun insert(user: User) {
        viewModelScope.launch(Dispatchers.IO) {
            mRepository.insert(user)
        }
    }

    fun insert(users: List<User>) {
        viewModelScope.launch(Dispatchers.IO) {
            mRepository.insertAll(users)
        }
    }

    fun delete(firstName: String) {
        viewModelScope.launch(Dispatchers.IO) {
            mRepository.delete(firstName)
        }
    }
}

@Suppress("UNCHECKED_CAST")
class UserViewModelFactory(private val repository: UserRepository) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(UserViewModel::class.java)) {
            return UserViewModel(repository) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}